lightning/src/ln/interactivetxs.rs

```diff
@@ -0,0 +1,667 @@
+	}
+
+	fn send_tx_add_output(mut self, serial_id: SerialId, output: TxOut) -> InteractiveTxStateMachine<Negotiating> {
+		self.context.outputs.insert(serial_id, output);
+		InteractiveTxStateMachine { context: self.context, state: Negotiating {} }
+	}
+
+	fn send_tx_remove_input(mut self, serial_id: SerialId) -> InteractiveTxStateMachine<Negotiating> {
+		self.context.inputs.remove(&serial_id);
+		InteractiveTxStateMachine { context: self.context, state: Negotiating {} }
+	}
+
+	fn send_tx_remove_output(mut self, serial_id: SerialId) -> InteractiveTxStateMachine<Negotiating> {
+		self.context.outputs.remove(&serial_id);
+		InteractiveTxStateMachine { context: self.context, state: Negotiating {} }
+	}
+
+	fn send_tx_abort(mut self) -> InteractiveTxStateMachine<NegotiationAborted> {
+		// A sending node:
+		// 	- MUST NOT have already transmitted tx_signatures
+		// 	- SHOULD forget the current negotiation and reset their state.
+		todo!();
+	}
+
+	fn receive_tx_abort(mut self) -> InteractiveTxStateMachine<NegotiationAborted> {
+		todo!();
+	}
+
+	fn is_current_transaction_state_able_to_complete(&self) -> Result<(), AbortReason> {
+		let tx_to_validate = Transaction {
+			version: self.context.base_tx.version,
+			lock_time: self.context.base_tx.lock_time,
+			input: self.context.inputs.values().cloned().collect(),
+			output: self.context.outputs.values().cloned().collect(),
+		};
+
+		// The receiving node:
+		// MUST fail the negotiation if:
+
+		// TODO: Verify this is the correct way to do this.
+		// - the peer's total input satoshis is less than their outputs
+		let get_output = |outpoint: &OutPoint| {
+			if outpoint.txid == tx_to_validate.txid() {
+				return tx_to_validate.output.get(outpoint.vout as usize).cloned()
+			} else {
+				None
+			}
+		};
+		if let Err(_) = tx_to_validate.verify(get_output) {
+			return Err(AbortReason::InvalidTransactionState)
+		};
+
+		// - there are more than 252 inputs
+		// - there are more than 252 outputs
+		if self.context.inputs.len() > MAX_INPUTS_OUTPUTS_COUNT ||
+			self.context.outputs.len() > MAX_INPUTS_OUTPUTS_COUNT {
+			return Err(AbortReason::ExceededNumberOfInputsOrOutputs)
+		}
+
+		if tx_to_validate.weight() > MAX_STANDARD_TX_WEIGHT {
+			return Err(AbortReason::TransactionTooLarge)
+		}
+
+		// TODO: Need to figure out how to do this
+		// if is the non-initiator:
+		// 	- the initiator's fees do not cover the common fields (version, segwit marker + flag,
+		// 		input count, output count, locktime)
+
+		return Ok(())
+	}
+
+	fn is_valid_counterparty_serial_id(&self, serial_id: SerialId) -> bool {
+		// A received `SerialId`'s parity must match the role of the counterparty.
+		self.context.holder_is_initiator == !serial_id.is_valid_for_initiator()
+	}
+}
+
+impl InteractiveTxStateMachine<TheirTxComplete> {
+	fn send_tx_complete(self) -> InteractiveTxStateMachine<NegotiationComplete> {
+		// TODO: Should we validate before transitioning states? If so, do we want to abort negotiation
```

Comment:
I presume, since [there isn't a prescription](https://github.com/lightning/bolts/blob/091397fc0798c4b2bdb36dbaa9f5b4b1d4a463e1/02-peer-protocol.md#requirements-3) that sending nodes need to fail if they have an invalid transaction state, we do not need to do a check similar to what we do for receiving `tx_complete` events (see `receive_tx_complete`)?